python_modules/dagster/dagster/_core/execution/plan/execute_step.py

```diff
@@ -474,9 +478,21 @@ def _get_output_asset_materializations(
         step_context.is_external_input_asset_records_loaded
         and asset_key in step_context.pipeline_def.asset_layer.asset_keys
     ):
-        tags = _build_logical_version_tags(asset_key, step_context)
-        logical_version = LogicalVersion(tags[LOGICAL_VERSION_TAG_KEY])
-        step_context.record_logical_version(asset_key, logical_version)
+        assert isinstance(output, Output)
+        code_version = _get_code_version(asset_key, step_context)
+        input_provenance_data = _get_input_provenance_data(asset_key, step_context)
+        logical_version = (
+            compute_logical_version(
+                code_version,
+                {k: meta["logical_version"] for k, meta in input_provenance_data.items()},
+            )
+            if output.logical_version is None
+            else output.logical_version
+        )
+        tags = _build_logical_version_tags(logical_version, code_version, input_provenance_data)
+        if not step_context.has_logical_version(asset_key):
```

Comment:
this is just asking for trouble. this has the potential to inject a silent failure condition where some code sets the logical version (e.g. the user does this in the user code themselves, which they can totally do) and breaks invariants in the system

Reply:
>this has the potential to inject a silent failure condition where some code sets the logical version (e.g. the user does this in the user code themselves, which they can totally do)

Hmm... it's technically possible for the user to do this, but the `StepExecutionContext` isn't part of the public API-- not sure I understand what separates this case from any other case of internal book-keeping machinery.